From 54d747f8182fb5b8e65b3c56be349421911b5a2c Mon Sep 17 00:00:00 2001
From: Jiali Chen <chenjiali@radxa.com>
Date: Tue, 30 Dec 2025 09:34:06 +0000
Subject: [PATCH 152/268] usb: typec: tcpm: support skipping VSafe0V check

On some hardware designs, the 5V VBUS output cannot be controlled or
disabled by the TCPM (e.g., fixed 5V supply or lack of a load switch).
Consequently, the VBUS voltage is unable to drop to the VSafe0V level
required by the standard state machine.

Currently, the TCPM blocks the transition from SRC_STARTUP to
SRC_ATTACHED until VSafe0V is detected. On hardware that cannot switch
off VBUS, this causes the port to hang or fail negotiation.

This patch adds a new device property "tcpm,skip-vsafe0v-check". When
set, the TCPM ignores the VSafe0V requirement and allows the state
machine to proceed, enabling support for hardware with uncontrollable
VBUS.

Signed-off-by: Jiali Chen <chenjiali@radxa.com>
---
 drivers/usb/typec/tcpm/tcpm.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/usb/typec/tcpm/tcpm.c
+++ b/drivers/usb/typec/tcpm/tcpm.c
@@ -388,6 +388,7 @@ struct tcpm_port {
 	 * Set to false when vbus is greater than VSAFE0V max.
 	 */
 	bool vbus_vsafe0v;
+	bool skip_vsafe0v_check;
 
 	bool vbus_never_low;
 	bool vbus_source;
@@ -4805,7 +4806,8 @@ static void run_state_machine(struct tcp
 		else if (tcpm_port_is_audio(port))
 			tcpm_set_state(port, AUDIO_ACC_ATTACHED,
 				       port->timings.cc_debounce_time);
-		else if (tcpm_port_is_source(port) && port->vbus_vsafe0v)
+		else if (tcpm_port_is_source(port) && 
+			(port->vbus_vsafe0v || port->skip_vsafe0v_check))
 			tcpm_set_state(port,
 				       tcpm_try_snk(port) ? SNK_TRY
 							  : SRC_ATTACHED,
@@ -7902,6 +7904,8 @@ struct tcpm_port *tcpm_register_port(str
 		goto out_unregister_pd;
 	}
 
+	port->skip_vsafe0v_check = device_property_read_bool(dev, "tcpm,skip-vsafe0v-check");
+
 	typec_port_register_altmodes(port->typec_port,
 				     &tcpm_altmode_ops, port,
 				     port->port_altmode, ALTMODE_DISCOVERY_MAX);
